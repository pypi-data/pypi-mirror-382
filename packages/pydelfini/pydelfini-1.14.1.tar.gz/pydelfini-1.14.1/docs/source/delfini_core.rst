Delfini Core library
===================

The :py:mod:`pydelfini.delfini_core` module is a set of low-level
endpoints and data models that have been automatically generated from
Delfini's OpenAPI spec. For every Delfini API endpoint, there is a
single API method in :py:mod:`pydelfini.delfini_core.api`, and for every
request and response body, there is a corresponding data model defined
in :py:mod:`pydelfini.delfini_core.models`.

Since this module is automatically generated, any missing or
insubstantial documentation should be addressed by adding the
appropriate fields to the main ``delfini-api.yaml`` file. See the
:ref:`Autogeneration` section for more information.

Usage
-----

First, create a client::

    from pydelfini.delfini_core import Client

    client = Client(base_url="https://delfini.bioteam.net/api/v1")

If the endpoints you're going to hit require authentication, login
using the :py:func:`pydelfini.delfini_core.login.Login` helper::

    from pydelfini.delfini_core import Client
    from pydelfini.delfini_core import Login

    client = Login(
        Client(base_url="https://delfini.bioteam.net/api/v1")
    ).with_password('myuser', 'mypassword')

Now call your endpoint and use your models::

    from pydelfini.delfini_core.models import CollectionsGetCollectionsCollectionList as CollectionList
    from pydelfini.delfini_core.api.collections import collections_get_collections
    from pydelfini.delfini_core.types import Response

    # the context manager is recommended in order to clean up any HTTP connection pool,
    # see https://www.python-httpx.org/advanced/
    with client as client:
        coll_list: CollectionList = collections_get_collections.sync(client=client)

        # or if you need more info (e.g. status_code)
        response: Response[CollectionList] = collections_get_collections.sync_detailed(client=client)

Or do the same thing with an async version::

    from pydelfini.delfini_core.models import CollectionsGetCollectionsCollectionList as CollectionList
    from pydelfini.delfini_core.api.collections import collections_get_collections
    from pydelfini.delfini_core.types import Response

    async with client as client:
        my_data: CollectionList = await collections_get_collections.asyncio(client=client)
        response: Response[CollectionList] = await collections_get_collections.asyncio_detailed(client=client)

By default, when you're calling an HTTPS API it will attempt to verify
that SSL is working correctly. Using certificate verification is
highly recommended most of the time, but sometimes you may need to
authenticate to a server (especially an internal server) using a
custom certificate bundle::

    client = AuthenticatedClient(
        base_url="https://delfini.local/api/v1",
        token="SuperSecretToken",
        verify_ssl="/path/to/certificate_bundle.pem",
    )

You can also disable certificate validation altogether, but beware
that **this is a security risk**::

    client = AuthenticatedClient(
        base_url="https://delfini.local/api/v1",
        token="SuperSecretToken",
        verify_ssl=False
    )

Things to know
--------------

1. Every path/method combo becomes a Python module with four functions:

    1. ``sync``: Blocking request that returns parsed data, if
       successful, or raises an exception otherwise.
    2. ``sync_detailed``: Blocking request that always returns a
       ``Request``, optionally with ``parsed`` set if the request was
       successful.
    3. ``asyncio``: Like ``sync`` but async instead of blocking
    4. ``asyncio_detailed``: Like ``sync_detailed`` but async instead of blocking

2. All path/query params, and bodies become method arguments.
3. If your endpoint had any tags on it, the first tag will be used as
   a module name for the function ("collections" in the above examples)
4. Any endpoint which did not have a tag will be in
   :py:mod:`pydelfini.delfini_core.api.default`

Advanced customizations
-----------------------

There are more settings on the generated ``Client`` class which let you
control more runtime behavior, check out the docstring on that class
for more info. You can also customize the underlying ``httpx.Client`` or
``httpx.AsyncClient`` (depending on your use-case)::

    from pydelfini.delfini_core import Client

    def log_request(request):
        print(f"Request event hook: {request.method} {request.url} - Waiting for response")

    def log_response(response):
        request = response.request
        print(f"Response event hook: {request.method} {request.url} - Status {response.status_code}")

    client = Client(
        base_url="https://delfini.bioteam.net/api/v1",
        httpx_args={"event_hooks": {"request": [log_request], "response": [log_response]}},
    )

    # Or get the underlying httpx client to modify directly with client.get_httpx_client()
    # or client.get_async_httpx_client()

You can even set the httpx client directly, but beware that this will
override any existing settings (e.g., base_url)::

    import httpx
    from pydelfini.delfini_core import Client

    client = Client(
        base_url="https://delfini.bioteam.net/api/v1",
    )
    # Note that base_url needs to be re-set, as would any shared cookies, headers, etc.
    client.set_httpx_client(
        httpx.Client(
            base_url="https://delfini.bioteam.net/api/v1", proxies="http://localhost:8030"
        )
    )

.. _Autogeneration:

Autogeneration
--------------

The core client code can be regenerated by running ``autogenerate.sh``
in the ``clients/pydelfini`` directory. This is automatically run by
the pre-commit hook system any time that ``delfini-api.yaml`` is
edited, so the client code should always remain in sync with the API
specification.

The ``autogenerate.sh`` script contains several templates and
workarounds for various bugs and strange behavior in
openapi-python-client.

Documentation
-------------

All documentation can be found under the
:py:mod:`pydelfini.delfini_core` API documentation section.

.. autosummary::
   :toctree: generated
   :template: custom-module-template.rst
   :recursive:

   pydelfini.delfini_core.api
   pydelfini.delfini_core.client
   pydelfini.delfini_core.errors
   pydelfini.delfini_core.login
   pydelfini.delfini_core.models
   pydelfini.delfini_core.paginator
   pydelfini.delfini_core.types


Updating Documentation
^^^^^^^^^^^^^^^^^^^^^^

Since the documentation is autogenerated from the code, and the code
is autogenerated from the original ``delfini-api.yaml`` API spec, any
updates to the :py:mod:`~pydelfini.delfini_core` documentation should be
made in ``delfini-api.yaml``. The mapping between fields in the source
OpenAPI document and the outputs in the generated code are:

* ``title:`` is essentially a rename for a given schema or endpoint.
  For example, if a schema is defined with the key
  ``dictionary_search_options:`` it will typically be autogenerated
  with the model title **DictionarySearchOptions**, but by providing
  ``title: Search Options for Dictionaries`` the model title will now
  be **SearchOptionsForDictionaries**. This should be used sparingly
  as it makes it harder to map back from the generated code to the
  original API definition.

* ``summary:`` is a short, one-line summary of a schema or endpoint.

* ``description:`` is a longer, multi-line description of the function
  of a schema or endpoint.
