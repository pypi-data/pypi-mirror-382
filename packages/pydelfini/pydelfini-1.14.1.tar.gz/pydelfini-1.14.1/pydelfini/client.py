"""Login and general Delfini operations"""
from collections.abc import Iterator
from urllib.parse import urlunparse

from .collections import DelfiniCollection
from .delfini_core import AuthenticatedClient as CoreAuthClient
from .delfini_core import Client as CoreClient
from .delfini_core import Login
from .delfini_core import Paginator
from .delfini_core.api.auth import auth_new_session
from .delfini_core.api.collections import collections_create_collection
from .delfini_core.api.collections import collections_get_collections
from .delfini_core.models import CollectionsGetCollectionsCollectionList
from .delfini_core.models import NewCollection
from .delfini_core.models import NewCollectionMetadata
from .delfini_core.types import UNSET


class DelfiniClient:
    """Main class for interacting with a Delfini instance."""

    def __init__(self, core: CoreAuthClient) -> None:
        self._client = core

    def __del__(self) -> None:
        if self._client._client:
            self._client._client.close()

    def get_collection_by_name(self, collection_name: str) -> DelfiniCollection:
        """Retrieve a single collection given its name.

        Raises:
            ValueError: if the collection could not be found or if more than one
                collection was found with the given name
        """
        collections = collections_get_collections.sync(
            client=self._client, name=collection_name
        )
        if len(collections.collections) == 1:
            return DelfiniCollection(collections.collections[0], self._client)

        raise ValueError(f"could not find collection: {collection_name}")

    def all_collections(self) -> Iterator[DelfiniCollection]:
        """Iterate over all collections visible to the user."""
        paginator = Paginator[CollectionsGetCollectionsCollectionList](
            collections_get_collections,
            self._client,
        )
        for page in paginator.paginate():
            for collection in page.collections:
                yield DelfiniCollection(collection, self._client)

    def new_collection(
        self, name: str, description: str, metadata: dict[str, str] = {}
    ) -> DelfiniCollection:
        """Create a new collection.

        Args:
            name: The short name of the collection
            description: A longer plain text description of the collection
            metadata: Optional key-value pairs of metadata on the collection

        """
        collection = collections_create_collection.sync(
            client=self._client,
            body=NewCollection(
                name=name,
                description=description,
                metadata=(
                    NewCollectionMetadata.from_dict(metadata) if metadata else UNSET
                ),
            ),
        )
        return DelfiniCollection(collection, self._client)


def make_base_url(host: str, scheme: str = "https") -> str:
    """Create a base URL for a Delfini instance.

    Args:
        host: The hostname of your Delfini instance, such as 'delfini.bioteam.net'.
        scheme: either 'https' or 'http'

    Returns:
        the Delfini API base URL

    """
    return urlunparse([scheme, host, "/api/v1", None, None, None])


def login(host: str, scheme: str = "https") -> DelfiniClient:
    """Login to a Delfini instance.

    This interactive login method will print a URL that you will need
    to visit in order to activate the login session. The function will
    wait to return until the session has successfully been activated.

    Args:
        host: The hostname of your Delfini instance, such as 'delfini.bioteam.net'.
        scheme: either 'https' or 'http'

    Returns:
        A :py:class:`DelfiniClient`

    """

    # get a session ID
    core = CoreClient(base_url=make_base_url(host, scheme))
    session_token = auth_new_session.sync(client=core)

    # construct the URL to activate the session
    activation_url = urlunparse(
        [
            scheme,
            host,
            f"/login/activate/{session_token.activation_code}",
            None,
            None,
            None,
        ]
    )

    print("To activate your session, visit the URL below:")
    print("  ", activation_url)
    print()
    print("Waiting for session activation...")

    # wait for the session to be activated
    authenticated = Login(core).with_session_id(session_token.session_id, wait=True)

    return DelfiniClient(authenticated)
