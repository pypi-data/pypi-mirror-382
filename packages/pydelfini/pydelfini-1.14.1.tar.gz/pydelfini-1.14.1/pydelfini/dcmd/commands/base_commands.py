import argparse
import json
import sys
import textwrap
import types
from typing import Any
from typing import Callable
from typing import Optional

import yaml
from pydelfini.client import DelfiniClient
from tabulate import tabulate


class BaseCommands:
    _func_args: dict[Callable[..., None], list[tuple[tuple[Any, ...], dict[str, Any]]]]

    def __init__(self, client: DelfiniClient, args: argparse.Namespace) -> None:
        self.client = client
        self.core = client._client
        self.args = args

    def _output(
        self,
        data: Any,
        header: Optional[str] = None,
        preferred_type: Optional[str] = None,
    ) -> None:
        output_type = self.args.output_type
        if output_type == "auto":
            output_type = preferred_type or "yaml"

        if output_type == "yaml":
            if header:
                print("#", header)
            print(yaml.dump(data))

        elif output_type == "json":
            print(json.dumps(data, indent=2))

        elif output_type == "table":
            try:
                print(tabulate(data, headers="keys"))
            except TypeError:
                print(
                    'This command does not support "table" output. Please choose'
                    " another output format."
                )
                sys.exit(1)
        else:
            raise ValueError(f"unknown output type: {output_type}")

    @classmethod
    def _add_to_parser(
        cls, subparsers: "argparse._SubParsersAction[argparse.ArgumentParser]"
    ) -> None:
        for method in dir(cls):
            if method.startswith("_"):
                continue

            action = getattr(cls, method)
            action_doc = action.__doc__ if action.__doc__ is not None else ""
            help_line = (
                action_doc.splitlines()[0].strip().rstrip(".") if action_doc else ""
            )
            parser = subparsers.add_parser(
                method.replace("_", "-"),
                help=help_line,
                description=(
                    help_line
                    + "\n\n"
                    + textwrap.dedent("\n".join(action_doc.split("\n")[1:]))
                ),
                formatter_class=argparse.RawDescriptionHelpFormatter,
            )

            if type(action) is type(object) and issubclass(action, BaseCommands):
                action._add_to_parser(parser.add_subparsers(title="subcommands"))

            elif isinstance(action, types.FunctionType):
                if hasattr(cls, "_func_args"):
                    for args, kwargs in cls._func_args.get(action, []):
                        parser.add_argument(*args, **kwargs)
                parser.set_defaults(cls=cls, func=action)

    @classmethod
    def _with_arg(cls, *args: Any, **kwargs: Any) -> Callable[..., Callable[..., None]]:
        def _inner(func: Callable[..., None]) -> Callable[..., None]:
            if not hasattr(cls, "_func_args"):
                cls._func_args = {}
            cls._func_args.setdefault(func, []).insert(0, (args, kwargs))
            return func

        return _inner
