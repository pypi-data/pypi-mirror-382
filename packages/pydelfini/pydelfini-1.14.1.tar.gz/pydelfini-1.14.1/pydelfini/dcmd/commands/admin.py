from pydelfini.delfini_core.api.admin import admin_grant_admin
from pydelfini.delfini_core.api.admin import admin_list_admins
from pydelfini.delfini_core.models import AdminGrantAdminAdminListRequest
from pydelfini.delfini_core.models import AdminGrants
from pydelfini.delfini_core.models import Operations

from .base_commands import BaseCommands
from .config import ConfigCommands


class AdminCommands(BaseCommands):
    """Administrative operations"""

    config = ConfigCommands

    def list_admins(self) -> None:
        """List system admins and privileges."""
        rv = admin_list_admins.sync(client=self.core)

        self._output(
            [
                {"subject_id": r.subject_id, "operations": ",".join(r.operations)}
                for r in rv.grants
            ],
            preferred_type="table",
        )

        if rv.operations:
            print()
            print("Operations:")
            for op in rv.operations:
                print("-", op)

    @BaseCommands._with_arg("subject_id")
    @BaseCommands._with_arg("-o", "--operation")
    def grant_admin(self) -> None:
        """Grant admin privilege to a subject ID."""
        admins = admin_list_admins.sync(client=self.core)

        if self.args.operation:
            new_operations = {Operations(self.args.operation)}
        else:
            new_operations = set(admins.operations or [])

        for grant in admins.grants:
            if grant.subject_id == self.args.subject_id:
                grant.operations = list(set(grant.operations) | new_operations)

                break
        else:
            admins.grants.append(
                AdminGrants(
                    subject_id=self.args.subject_id, operations=list(new_operations)
                )
            )

        rv = admin_grant_admin.sync(
            body=AdminGrantAdminAdminListRequest(grants=admins.grants), client=self.core
        )
        assert rv

        print("Updated", self.args.subject_id, " -- now admins are:")
        self.list_admins()

    @BaseCommands._with_arg("subject_id")
    def revoke_admin(self) -> None:
        """Revoke all admin privileges from a user."""
        admins = admin_list_admins.sync(client=self.core)

        new_grants = [g for g in admins.grants if g.subject_id != self.args.subject_id]
        if new_grants == admins.grants:
            raise ValueError(f"{self.args.subject_id} is not listed in admins")

        admin_grant_admin.sync(
            body=AdminGrantAdminAdminListRequest(grants=new_grants), client=self.core
        )
