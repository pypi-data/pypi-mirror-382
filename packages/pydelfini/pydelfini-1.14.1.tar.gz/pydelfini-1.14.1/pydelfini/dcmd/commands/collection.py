import os.path
import shutil

from pydelfini.delfini_core.api.collections import collections_create_collection
from pydelfini.delfini_core.api.collections import collections_delete_collection_version
from pydelfini.delfini_core.api.collections import collections_get_collections
from pydelfini.delfini_core.models import NewCollection

from .base_commands import BaseCommands


class CollectionCommands(BaseCommands):
    """Operations on collections"""

    def list(self) -> None:
        """List all visible collections."""
        collections = collections_get_collections.sync(client=self.core)

        self._output(collections.to_dict())

    @BaseCommands._with_arg("name", help="Collection name, required")
    @BaseCommands._with_arg(
        "-d", "--description", help="Collection description", default=""
    )
    def new(self) -> None:
        """Create a new collection."""
        collection = collections_create_collection.sync(
            body=NewCollection(
                name=self.args.name,
                description=self.args.description,
            ),
            client=self.core,
        )

        self._output(collection.to_dict())

    @BaseCommands._with_arg("collection_id", help="Collection ID")
    @BaseCommands._with_arg(
        "-v", "--version_id", help="Collection version ID", default="LIVE"
    )
    def delete(self) -> None:
        """Delete a collection."""
        rv = collections_delete_collection_version.sync(
            collection_id=self.args.collection_id,
            version_id=self.args.version_id,
            client=self.core,
        )
        assert rv

    @BaseCommands._with_arg("collection_name", help="Collection name")
    @BaseCommands._with_arg("filename", help="File to upload")
    def upload(self) -> None:
        """Upload a file to a collection."""
        collection = self.client.get_collection_by_name(self.args.collection_name)

        with open(self.args.filename, "rb") as src, collection.open(
            os.path.basename(self.args.filename), "wb"
        ) as dest:
            shutil.copyfileobj(src, dest)

    @BaseCommands._with_arg("collection_name", help="Collection name")
    @BaseCommands._with_arg("filename", help="File to download")
    def download(self) -> None:
        """Download an item from a collection."""
        collection = self.client.get_collection_by_name(self.args.collection_name)

        with collection.open(self.args.filename, "rb") as src, open(
            os.path.basename(self.args.filename), "wb"
        ) as dest:
            shutil.copyfileobj(src, dest)
