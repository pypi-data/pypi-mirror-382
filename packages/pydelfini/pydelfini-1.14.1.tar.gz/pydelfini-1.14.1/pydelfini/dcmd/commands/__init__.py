from .accounts import AccountsCommands
from .admin import AdminCommands
from .auth import AuthCommands
from .base_commands import BaseCommands
from .cde import CdeCommands
from .collection import CollectionCommands
from .groups import GroupsCommands
from .pdd import PddCommands
from .tasks import TasksCommands
from .users import UsersCommands


class Commands(BaseCommands):
    accounts = AccountsCommands
    admin = AdminCommands
    auth = AuthCommands
    cdeset = CdeCommands
    collection = CollectionCommands
    groups = GroupsCommands
    pdd = PddCommands
    tasks = TasksCommands
    users = UsersCommands
