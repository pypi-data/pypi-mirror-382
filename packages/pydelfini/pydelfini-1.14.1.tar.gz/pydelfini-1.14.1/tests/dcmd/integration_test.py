import argparse
import ast
import os
import random
import re
import shlex
from pathlib import Path
from urllib.request import urlopen

import pytest
from pydelfini.client import make_base_url
from pydelfini.dcmd import cli
from pydelfini.dcmd.commands import Commands as PcmdCommands
from pydelfini.delfini_core import Client as CoreClient
from pydelfini.delfini_core.api.user import user_create_user
from pydelfini.delfini_core.login import Login
from pydelfini.delfini_core.login import LoginError
from pydelfini.delfini_core.login import to_token_file
from pydelfini.delfini_core.models import NewUser
from yaml import dump as yaml_dump
from yaml import safe_load


# The tests in this test suite are designed to be run against a local
# development instance of Delfini hosted at localhost:8170, similar to
# the Cypress tests in `delfini-nuxt`. In order to be able to run these
# tests, be sure to have the local development instance running in the
# background with the command:
#
#    flask run --debug
#


@pytest.fixture
def local_user(tmp_path):
    # try to log in with the credentials
    username, password = "dcmd.tester", "insecure"
    base_url = make_base_url("localhost:8170", "http")
    try:
        authenticated = Login(CoreClient(base_url=base_url)).with_password(
            username, password
        )
    except LoginError:
        print("Login failed, trying to create new user...")
        user_create_user.sync(
            body=NewUser(
                given_name="dcmd",
                last_name="tester",
                user_email="john.doe@example.com",
                user_name=username,
                password=password,
            ),
            client=CoreClient(base_url=base_url),
        )
        authenticated = Login(CoreClient(base_url=base_url)).with_password(
            username, password
        )

    token_file = tmp_path / "token"
    to_token_file(authenticated, token_file)

    yield str(token_file)


def random_id():
    return f"rr{random.randint(0, 1048576):05x}rr"


def mask(rv):
    # mask UUIDs
    rv = re.sub(
        r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}",
        "00000000-0000-0000-0000-000000000000",
        rv,
    )
    # mask timestamps
    rv = re.sub(
        r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.[0-9+:]*)?",
        "1970-01-02T03:04:05.000000",
        rv,
    )
    # mask task worker
    rv = re.sub(r"worker: .*", "worker: dcmd-test-mask", rv)
    # mask random ID
    rv = re.sub(r"rr[0-9a-f]{5}rr", "random-id", rv)
    # mask bundle ID
    rv = re.sub(r"b-[a-zA-Z0-9]{8}", "b-00000000", rv)

    return rv


@pytest.fixture
def dcmd(local_user, capsys):
    def _cmd(args, masked=True):
        if isinstance(args, str):
            args = shlex.split(args)
        cli.main(["-H", "localhost:8170", "--insecure", "-k", local_user] + args)

        outv = capsys.readouterr()
        if not outv.err:
            rv = outv.out
            if masked:
                rv = mask(rv)
            return rv

        return outv

    yield _cmd


@pytest.fixture
def test_target_user(dcmd):
    response = user_create_user.sync(
        body=NewUser(
            given_name="dcmd",
            last_name="test target",
            user_email=f"{random_id()}@example.com",
            user_name="dcmd.test.target",
            password=random_id(),
        ),
        client=CoreClient(base_url=make_base_url("localhost:8170", "http")),
    )

    yield response.subject_id

    dcmd("users delete dcmd.test.target")


@pytest.fixture
def scratch_collection(dcmd):
    collection_name = f"dcmd-test.{random_id()}"
    rv = safe_load(dcmd(f"collection new {collection_name}", masked=False))

    yield collection_name, rv["id"]

    dcmd(f"collection delete {rv['id']}")


def test_dcmd_auth_whoami(dcmd, snapshot):
    dcmd('auth set-metadata x-tos \'{"toc":3,"coc":6,"priv":3,"com":1}\'')

    assert dcmd("auth whoami") == snapshot


# NOTE: If the following tests are failing with "Unexpected status
# code: 403", you should grant the `dcmd.tester` user full admin
# access to your local development instance with the command:
#
#   flask admin grant dcmd.tester
#


def test_dcmd_tasks(dcmd, snapshot):
    stats = safe_load(dcmd("tasks stats"))
    assert stats.keys() == {"queued", "results", "schedules"}

    schedules = safe_load(dcmd("tasks schedules"))["schedules"]
    assert len(schedules) >= 2

    results = safe_load(dcmd("tasks results --idemkey tasks-backends-sqlite-vacuum"))[
        "results"
    ]
    if len(results) > 0:
        assert results[0] == snapshot


def test_dcmd_users(dcmd, test_target_user):
    rv = dcmd("users list")
    assert set(rv.splitlines()[0].split()) == {
        "account_id",
        "created_on",
        "metadata",
        "subject_id",
        "user_name",
        "visibility_level",
        "given_name",
        "groups",
        "last_name",
        "user_email",
    }
    assert "dcmd.test.target" in rv

    # disable and enable
    rv = safe_load(dcmd("users disable dcmd.test.target"))
    assert rv["disabled_on"] is not None

    rv = safe_load(dcmd("users enable dcmd.test.target"))
    assert rv["disabled_on"] is None


def test_dcmd_groups(dcmd, test_target_user, snapshot):
    group_name = f"test-dcmd-group.{random_id()}"
    new_group = dcmd(f"groups new {group_name}", masked=False)
    group_id = safe_load(new_group)["id"]
    assert mask(new_group) == snapshot

    groups = safe_load(dcmd("groups list"))["groups"]
    assert safe_load(mask(new_group)) in groups

    dcmd(f"groups join {group_id} {test_target_user}")

    assert len(safe_load(dcmd(f"groups list-members {group_id}"))["members"]) == 2

    dcmd(f"groups update --name new-test-group {group_id}")

    dcmd(f"groups delete {group_id}")


def test_dcmd_accounts(dcmd, snapshot, tmp_path, scratch_collection):
    collection_name, collection_id = scratch_collection

    account_name = f"test-dcmd-account.{random_id()}"
    new_account = dcmd(f"accounts new {account_name} --description heyyo", masked=False)
    account_id = safe_load(new_account)["id"]
    assert mask(new_account) == snapshot

    accounts = safe_load(dcmd("accounts list", masked=False))["accounts"]
    assert safe_load(new_account) in accounts

    # update space
    index_file = tmp_path / "index.md"
    with open(index_file, "w") as fp:
        fp.write("# Hello world!\n")
    dcmd(f"collection upload {collection_name} {index_file}")
    dcmd(f"accounts update-space {account_id} {collection_id}")
    response = urlopen(
        f"http://localhost:8170/api/v1/account/{account_id}/pages/index.md"
    )
    assert response.code == 200

    dcmd(f"accounts delete {account_id}")

    accounts = safe_load(dcmd("accounts list", masked=False))["accounts"]
    assert safe_load(new_account) not in accounts


def test_dcmd_admin_grant(dcmd, test_target_user):
    dcmd(f"admin grant-admin -o group {test_target_user}")

    rv = dcmd("admin list-admins", masked=False)
    assert test_target_user in rv

    dcmd(f"admin revoke-admin {test_target_user}")

    rv = dcmd("admin list-admins", masked=False)
    assert test_target_user not in rv


def test_dcmd_admin_config(dcmd, tmp_path, capsys):
    rv = safe_load(dcmd("admin config get"))
    assert "datastores" in rv
    assert "authentication" in rv

    dcmd(
        "admin config set datastores.dcmd_test"
        + f' \'{{"enabled": false, "path": "{tmp_path}", "type": "local"}}\''
    )

    rv = safe_load(dcmd("admin config get", masked=False))
    assert "dcmd_test" in rv["datastores"]

    newdir = tmp_path / "new"
    rv["datastores"]["dcmd_test"]["path"] = str(newdir)

    config_fn = tmp_path / "config.yaml"
    with open(config_fn, "w") as fp:
        yaml_dump(rv, fp)

    dcmd(f"admin config put {config_fn}")

    rv = safe_load(dcmd("admin config get"))
    assert "dcmd_test" in rv["datastores"]
    assert rv["datastores"]["dcmd_test"]["path"] == str(newdir)

    dcmd("admin config delete datastores.dcmd_test")

    rv = safe_load(dcmd("admin config get"))
    assert "dcmd_test" not in rv["datastores"]


def test_dcmd_collection(dcmd, tmp_path, scratch_collection):
    collection_name, collection_id = scratch_collection
    rv = dcmd("collection list", masked=False)
    assert collection_name in rv
    assert collection_id in rv

    test_file = tmp_path / "test.txt"
    with open(test_file, "w") as fp:
        fp.write("hello world this is a test\n")

    dcmd(f"collection upload {collection_name} {test_file}")

    test_file.unlink()

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    dcmd(f"collection download {collection_name} test.txt")
    os.chdir(old_cwd)

    assert test_file.exists()


def test_dcmd_pdd(dcmd, scratch_collection, snapshot):
    collection_name, collection_id = scratch_collection
    common_pdd_fn = Path(__file__).parent / "common_pdd.json"

    rv = dcmd(f"pdd upload {collection_id} {common_pdd_fn}", masked=False)
    assert mask(rv) == snapshot
    safe_load(rv)["id"]

    assert dcmd(f"pdd list {collection_name} common_pdd.json") == snapshot

    dcmd(f"pdd set-order {collection_name} common_pdd.json 6fr70ICV O8zPEsiM")

    assert dcmd(f"pdd list {collection_name} common_pdd.json") == snapshot

    assert (
        dcmd(f"pdd bundle {collection_name} common_pdd.json MyBundle O8zPEsiM 6fr70ICV")
        == snapshot
    )

    assert dcmd(f"pdd list {collection_name} common_pdd.json") == snapshot


def test_dcmd_cdeset(dcmd, scratch_collection):
    collection_name, collection_id = scratch_collection
    common_pdd_fn = Path(__file__).parent / "common_pdd.json"

    rv = dcmd(f"pdd upload {collection_id} {common_pdd_fn}", masked=False)
    item_id = safe_load(rv)["id"]

    cdeset_name = f"dcmd.{random_id()}"
    dcmd(f"cdeset new {cdeset_name} 'dcmd test cde set'")

    rv = dcmd("cdeset list", masked=False)
    assert cdeset_name in rv

    dcmd(f"cdeset copy-from-pdd {cdeset_name} {collection_id} LIVE {item_id}")

    rv = safe_load(dcmd("cdeset list", masked=False))
    cdeset = {"f": c for c in rv["cdesets"] if c["name"] == cdeset_name}.get("f")
    assert cdeset["num_cdes"] == 2
    assert cdeset["description"] == "dcmd test cde set"

    dcmd(f"cdeset delete {cdeset_name}")


###
### Coverage Verification
###


class CoverageScanner(ast.NodeVisitor):
    def __init__(self):
        super().__init__()
        self.actions = {}

    def register_actions(self, parser, actlevel=None):
        if actlevel is None:
            actlevel = self.actions

        actions = parser._get_positional_actions()
        if actions and actions[0].choices:
            for arg, subparser in actions[0].choices.items():
                self.register_actions(subparser, actlevel.setdefault(arg, {}))

    def coverage_needed(self):
        needed = set()

        def _walk(d, s=None):
            if s is None:
                s = []
            if "SEEN" in d:
                return
            if not d:
                needed.add(" ".join(s))
            for k in d:
                _walk(d[k], s + [k])

        _walk(self.actions)
        return needed

    def visit_Call(self, node):
        if not isinstance(node.func, ast.Name):
            return self.generic_visit(node)
        if node.func.id != "dcmd":
            return self.generic_visit(node)
        if not node.args:
            return self.generic_visit(node)

        def _resolve(arg):
            if isinstance(arg, ast.Constant):
                return arg.value
            elif isinstance(arg, ast.JoinedStr):
                return _resolve(arg.values[0])
            elif isinstance(arg, ast.BinOp):
                return _resolve(arg.left)
            else:
                print(ast.dump(node, indent=2))
                raise Exception(
                    f"unexpected dcmd call on line {node.lineno}: {ast.unparse(node)}"
                )

        dcmd_arg = _resolve(node.args[0])

        actlevel = self.actions
        for word in shlex.split(dcmd_arg):
            if word in actlevel:
                actlevel = actlevel[word]

        if not actlevel:
            actlevel["SEEN"] = True


def test_dcmd_coverage():
    # This checks all of the dcmd() calls in the AST of this file and
    # matches them up with commands defined in the dcmd command set.
    # Any commands not called will show up in
    # `scanner.coverage_needed()`.

    argparser = argparse.ArgumentParser()
    PcmdCommands._add_to_parser(
        argparser.add_subparsers(title="commands", required=True)
    )
    scanner = CoverageScanner()
    scanner.register_actions(argparser)

    test_ast = ast.parse(open(__file__).read(), filename=__file__)
    scanner.visit(test_ast)

    assert scanner.coverage_needed() == set()
