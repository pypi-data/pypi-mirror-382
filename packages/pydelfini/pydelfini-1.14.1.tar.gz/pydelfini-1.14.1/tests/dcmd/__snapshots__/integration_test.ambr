# serializer version: 1
# name: test_dcmd_accounts
  '''
  createdOn: '1970-01-02T03:04:05.000000'
  id: a:00000000-0000-0000-0000-000000000000
  isPersonal: false
  membersGroupId: g:00000000-0000-0000-0000-000000000000
  metadata:
    x-description: heyyo
  name: test-dcmd-account.random-id
  visibilityLevel: public
  
  
  '''
# ---
# name: test_dcmd_auth_whoami
  '''
  expires: '1970-01-02T03:04:05.000000'
  user:
    account_id: a:00000000-0000-0000-0000-000000000000
    email: john.doe@example.com
    has_operations:
    - account
    - admin.cde
    - admin.config
    - admin.grantadmin
    - admin.metadata
    - admin.metrics
    - admin.tasks
    - admin.users
    - group
    identity:
      fqda: dcmd.tester@localhost
      is_local_user: true
      primary_id: 00000000-0000-0000-0000-000000000000
      user_name: dcmd.tester
    image: null
    metadata:
      x-tos: '{"toc":3,"coc":6,"priv":3,"com":1}'
    name: dcmd tester
  
  
  '''
# ---
# name: test_dcmd_groups
  '''
  accountLinked: false
  controlledAccess: false
  createdOn: '1970-01-02T03:04:05.000000'
  id: g:00000000-0000-0000-0000-000000000000
  metadata: {}
  name: test-dcmd-group.random-id
  visibilityLevel: unlisted
  
  
  '''
# ---
# name: test_dcmd_pdd
  '''
  id: 00000000-0000-0000-0000-000000000000
  
  '''
# ---
# name: test_dcmd_pdd.1
  '''
  [O8zPEsiM] Age
  [6fr70ICV] Zip Code
  
  '''
# ---
# name: test_dcmd_pdd.2
  '''
  [6fr70ICV] Zip Code
  [O8zPEsiM] Age
  
  '''
# ---
# name: test_dcmd_pdd.3
  '''
  New bundle name: MyBundle
  Element IDs: ['O8zPEsiM', '6fr70ICV']
  
  New bundle ID: b-00000000
  
  '''
# ---
# name: test_dcmd_pdd.4
  '''
  MyBundle
     [O8zPEsiM] Age
     [6fr70ICV] Zip Code
  
  '''
# ---
# name: test_dcmd_tasks
  dict({
    'data': None,
    'errors': list([
    ]),
    'event_time': '1970-01-02T03:04:05.000000',
    'id': '00000000-0000-0000-0000-000000000000',
    'start_time': '1970-01-02T03:04:05.000000',
    'status': 'OK',
    'task_id': '00000000-0000-0000-0000-000000000000',
    'task_idemkey': 'tasks-backends-sqlite-vacuum',
    'worker': 'dcmd-test-mask',
  })
# ---
